kernel/arch/dreamcast/include/arch/perfctr.h

```diff
@@ -0,0 +1,271 @@
+/** \brief  Clear a performance counter.
+
+    This function clears a performance counter. It resets its count to zero.
+    This function stops the counter before clearing it because you cant clear 
+    a running counter.
+
+    \param  counter           The counter to clear (i.e, \ref PRFC0 or PRFC1).
+*/
+void perf_cntr_clear(perf_cntr_t counter);
+
+/** \brief  Obtain the count of a performance counter.
+
+    This function simply returns the count of the counter.
+
+    \param  counter         The counter to read (i.e, \ref PRFC0 or PRFC1).
+    
+    \return                 The counter's count.
+*/
+uint64_t perf_cntr_count(perf_cntr_t counter);
+
+/** \defgroup perf_counters_timer Timer
+    \brief    High-resolution performance counter-based timer API
+
+    This API allows for using the performance counters as high-resolution
+    general-purpose interval timer with 5ns ticks. It does this by configuring
+    \ref PRFC0 in \ref PMCR_ELAPSED_TIME_MODE.
+
+    \note
+    This is enabled by default. To use \ref PRFC0 for something else, either
+    use perf_cntr_timer_disable() or perf_cntr_start() to reconfigure it for
+    something else. When disabled, the timer will simply fall through to use
+    timer_ns_gettime64() from the timer driver, decreasing the resolution of
+    each tick to 80ns.
+
+    \warning
+    The performance counter timer is only counting *active* CPU cycles. This
+    means that when KOS's thread scheduler uses the "sleep" instruction, 
+    putting the CPU to sleep, these counters cease to record elapsed time. 
+    Because of this, they should only be used to measure small deltas that
+    are not across frames.
+
+    \sa timers
+
+    @{
+*/
+
+/** \brief  Enable the nanosecond timer.
+
+    This function enables the performance counter used for the perf_cntr_timer_ns()
+    function. 
+    
+    \note
+    This is on by default. The function uses \ref PRFC0 to do the work.
+
+    \warning
+    The 
+*/
+void perf_cntr_timer_enable(void);
+
+/** \brief  Disable the nanosecond timer.
+
+    This function disables the performance counter used for the
+    perf_cntr_timer_ns() function. 
+
+    \note
+    Generally, you will not want to do this, unless you have some need to use 
+    the counter \ref PRFC0 for something else.
+*/
+void perf_cntr_timer_disable(void);
+
+/** \brief Check whether the nanosecond timer is enabled.
+    
+    Queries the configuration of \ref PRFC0 to check whether it's 
+    currently acting as the nanosecond timer.
+
+    \note
+    Even when it's not, perf_cntr_timer_ns() will still gracefully fall-through
+    to using the timer_ns_gettime64(), which decreases the resolution of each
+    tick to 80ns.
+
+    \retval true    The nanosecond timer is configured and running
+    \retval false   The nanosecond timer is not configured and/or isn't
+                    running
+*/
+bool perf_cntr_timer_enabled(void);
+
+/** \brief  Get the current uptime of the system (in nanoseconds).
```

Comment:
Since we found out that the sleep instruction messes with the timing wouldn't the documentation on the function be incorrect?

Reply:
Maybe add a warning to this function that it doesnt include the sleep time?

Reply:
SHIT. I did that in the topic, but you're right, it needs to be done there as well.